Add plugin template registration API

Aljullu opened the pull request:
This PR includes the changes that need to be backported from https://github.com/WordPress/gutenberg/pull/61577 and https://github.com/WordPress/gutenberg/pull/64610 into WC core.

Trac ticket: https://core.trac.wordpress.org/ticket/61804

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Aljullu committed:
Add plugin template registration API

Aljullu committed:
Rename 'templates' to 'block templates' in several APIs

Aljullu committed:
Simplify get_block_file_template() logic

Aljullu commented:
@youknowriad I've tested the PHP APIs in this PR and they are working as expected. Unfortunately, I didn't find a way to reproduce all the testing steps from https://github.com/WordPress/gutenberg/pull/61577. This PR doesn't include the JS changes from the original PR so template revert/removal doesn't work. Is it expected that this kind of PRs can't be tested right away, or am I missing something?

youknowriad commented:
@Aljullu yes, that's fine, it means this PR can only be merged after the package update on Core.

Aljullu committed:
Change default author text from theme slug to the plugin slug

Aljullu committed:
Remove unnecessary usage of empty

Aljullu committed:
DRY when validating registry input

Aljullu committed:
Fix typo in error message

Aljullu committed:
Fix typo in variable name

Aljullu committed:
Fix wrong register method signature

Aljullu committed:
Make plugin-registered templates overriden by themes to fall back to plugin-registered title and description

peterwilsoncc commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1157,6 +1183,25 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 		}
+
+		if ( 'wp_template' === $template_type ) {
+			/*
+			 * Add templates registered in the template registry. Filtering out the ones which have a theme file.
+			 */
```
```suggestion
			// Add templates registered in the template registry. Filtering out the ones which have a theme file.
```

peterwilsoncc commented on src/wp-includes/block-template.php:
```diff
@@ -358,3 +358,34 @@ function _resolve_template_for_new_post( $wp_query ) {
 }
+
+/**
+ * Register a template.
```
* `@since` annotation. 
* A further explanation of the kind of template being registered would be good to avoid confusion with template files in classic themes, etc

Aljullu replied:
> * @since annotation.

Added in 8188a0fb1fc69a10bc22e712aca9a53ff3af9492.

> A further explanation of the kind of template being registered would be good to avoid confusion with template files in classic themes, etc

I updated the comments to mention 'block templates' instead of simply 'templates' in b3e73e0be63ad01cf022dc90f99e5620b2829cc1. Do you think that's ok?

peterwilsoncc commented on src/wp-includes/block-template.php:
```diff
@@ -358,3 +358,34 @@ function _resolve_template_for_new_post( $wp_query ) {
+ *     @type string        $content               Optional. Default content of the template that will be used when the
+ *                                                template is rendered or edited in the editor.
+ *     @type string[]      $post_types            Optional. Array of post types to which the template should be available.
+ *     @type string        $plugin                Uri of the plugin that registers the template.
```
Reviewing the tests this looks optional.

```suggestion
 *     @type string        $plugin                Optional. Uri of the plugin that registers the template.
```

Also, is it the URI or the slug?

Aljullu replied:
Good catch! Done in b1f9ec20d3406b4060a52b39a86eded9f56aacb6.

peterwilsoncc commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+		$error_code    = '';
+
+		if ( ! is_string( $template_name ) ) {
+			$error_message = __( 'Template names must be strings.', 'gutenberg' );
```
Remove gutenberg domain, :1234: applies throughout.

```suggestion
			$error_message = __( 'Template names must be strings.' );
```

Aljullu replied:
Fixed in cf747f8c88208607a5f8cd836a3377541cb63b38.

peterwilsoncc reviewed (dismissed):
Added a few notes for the wp-dev version.

Aljullu committed:
Remove 'gutenberg' domain from translation strings

Aljullu committed:
Make  variable as optional and clarify it's the plugin slug

Aljullu committed:
Add since tags to new functions

Aljullu committed:
Clarify registering functions refer to block templates

Aljullu committed:
Update src/wp-includes/block-template-utils.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

Aljullu commented:
Thanks for the review @peterwilsoncc! I have applied all your feedback.

anton-vlasenko commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -1134,6 +1173,12 @@ public function get_item_schema() {
 			);
+			$schema['properties']['plugin']    = array(
+				'type'        => 'string',
+				'description' => __( 'Plugin that registered the template.', 'gutenberg' ),
```
```suggestion
				'description' => __( 'Plugin that registered the template.' ),
```

anton-vlasenko commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+/**
+ * Templates registry functions.
+ *
+ * @package Gutenberg
```
```suggestion
 * @package WordPress
```

Aljullu replied:
Thanks, @anton-vlasenko! I applied both suggestions.

anton-vlasenko reviewed (commented):
I haven't had a chance to do a thorough review, but this caught my attention.
Thank you for considering my notes.

azaozz commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -865,9 +872,41 @@ private static function get_wp_templates_author_text_field( $template_object ) {
+				if ( isset( $template_object->plugin ) ) {
+					$plugins = wp_get_active_and_valid_plugins();
+
+					foreach ( $plugins as $plugin_file ) {
+						$plugin_basename = plugin_basename( $plugin_file );
+						// Split basename by '/' to get the plugin slug.
+						list( $plugin_slug, ) = explode( '/', $plugin_basename );
+
+						if ( $plugin_slug === $template_object->plugin ) {
+							$plugin_data = get_plugin_data( $plugin_file );
+
+							if ( ! empty( $plugin_data['Name'] ) ) {
+								return $plugin_data['Name'];
+							}
+
+							break;
+						}
+					}
```
This seems somewhat slow, mostly the `get_plugin_data()`. May be missing something but seems this is supposed to run on every page load for all templates that have a "plugin" in the settings. If that's the case, perhaps some caching would be nice here? 

Or even better, is the readable plugin author name always necessary? Would it make sense to retrieve it only when it's needed, otherwise just return whatever is in `$template_object->plugin`?

peterwilsoncc replied:
I agree with Ozz that it would be good to avoid the performance impact here, the function is in the admin folder section because it's very slow.

Aljullu replied:
Thanks for the comments!

> Or even better, is the readable plugin author name always necessary? Would it make sense to retrieve it only when it's needed, otherwise just return whatever is in $template_object->plugin?

This code runs only when querying the templates endpoint. That means the code does not run on the frontend or on most `wp-admin` pages. It runs in the Site Editor, where the template author text is displayed in several UI elements, so I'm not sure how it can be optimized in that sense.

When making a call to the templates endpoint, this specific code path can be deactivated if the `author_text` attribute is not requested. That could be done in some specific cases, like when loading templates for the _Swap template_ popup in the Post Editor, as AFAIK we don't display the author text anywhere in the UI, but I might be wrong. :thinking: 

> If that's the case, perhaps some caching would be nice here?

I agree that caching could help. Since this is my first PR of this size in WP core, I'd appreciate any specific recommendations on implementing caching in this context. My initial thought is to store a transient that maps plugin slugs to plugin names, which would be invalidated whenever a plugin is activated, deactivated, or updated. But wondering if you have any better suggestions. :slightly_smiling_face:

peterwilsoncc replied:
> My initial thought is to store a transient that maps plugin slugs to plugin names, which would be invalidated whenever a plugin is activated, deactivated, or updated.

The various hooks for when a plugin is updated only run if the plugin is updated via the Dashboard. In the case of sites deployed by git, rsync or good old FTP this isn't the case which makes caching these functions persistently via a transient or the object cache unworkable.

azaozz replied:
> This code runs only when querying the templates endpoint. That means the code does not run on the frontend

Thanks for the clarification. As peterwilsoncc pointed out above when plugins are updated by FTP (uploading) or from version control, WP doesn't know it unless an admin visits the Plugins screen. This makes caching of the plugins meta somewhat impractical as that cache will always be stale in some cases/on some sites. 

On the other hand mapping the plugin's name to the plugin's slug may be acceptable even if a bit stale. It is very rare that a plugin is renamed, so probably can use a transient that expires every week or so? That may be an acceptable compromise as this information doesn't seem critical.

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -865,9 +872,41 @@ private static function get_wp_templates_author_text_field( $template_object ) {
+				 * Fall back to the theme name if the plugin is not defined. That's needed to keep backwards
+				 * compatibility with templates that were registered before the plugin attribute was added.
+				 */
+				$plugins         = get_plugins();
```
As above with `get_plugins()`, it gets a file listing of `wp-content/plugins/*`, filters for PHP files and opens each of them to look for a plugin header. In the case of Jetpack, this is 38 files.

`get_plugins()` is cached but the `plugins` group isn't persistent so the code runs on each http request calling the function.

Aljullu replied:
In cases where a plugin is using the API and setting the namespace to match its slug, we will have already returned in this function, so this code would never run.

The main reason to keep it is for backwards compatibility, as this code already exists in WP 6.6 and some plugins rely on it: they don't use the new template registration API but they still set the template source to `plugin`.

Aljullu committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

Aljullu committed:
Update src/wp-includes/class-wp-block-templates-registry.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

peterwilsoncc commented on tests/phpunit/tests/block-template.php:
```diff
@@ -431,6 +431,33 @@ public function test_get_block_templates_paths_dir_doesnt_exists() {
 		$this->assertSame( array(), $template_paths );
 	}
 
+	public function test_get_block_templates_from_registry() {
```
Docblock:

:1234: This applies to each of the new tests.

* `@ticket` annotation
* `@covers` annotations
* Description of the test

Aljullu replied:
Thanks for the feedback, I added docblocks to all tests in f6a3dfa689ecb1afb8c2cef61a69a3dad6fedbe9.

peterwilsoncc commented on tests/phpunit/tests/class-wp-templates-registry-test.php:
```diff
@@ -0,0 +1,192 @@
+ *
+ * @covers WP_Block_Templates_Registry
+ */
+class WP_Block_Templates_Registry_Test extends WP_UnitTestCase {
```
Rename and relocate file to `block-templates/WpBlockTemplatesRegistry.php` too

```suggestion
class Tests_Block_Templates_WpBlockTemplatesRegistry extends WP_UnitTestCase {
```

Aljullu replied:
Done in fa9e6ad681636255c2ee3f48eafc3aa02f84b921.

peterwilsoncc reviewed (commented):
I've added a few notes for the tests. As noted inline, the test dockblocks will need to be added to each of the tests to allow future devs to filter by ticket number (use the trac ticket reference rather than the gutenberg repo's).

Although this needs a little work before landing in the final release, I am a little tempted to merge the code in it's current form while performance is figured out so we can merge in the inital package update. What do you think @azaozz @noisysocks?

anton-vlasenko commented on src/wp-includes/block-template-utils.php:
```diff
@@ -592,6 +592,19 @@ function _build_block_template_result_from_file( $template_file, $template_type
+	if ( 'wp_template' === $template_type ) {
+		$registered_template = WP_Block_Templates_Registry::get_instance()->get_by_slug( $template_file['slug'] );
+		if ( $registered_template ) {
+			$template->plugin      = $registered_template->plugin;
```
The `WP_Block_Template::$plugin` property must be declared on the class if you need to use it.
PHP 8.2 and above do not support dynamic class properties: https://3v4l.org/BZI1O#v8.2.23.  
Related proposal: https://core.trac.wordpress.org/ticket/56034

Aljullu replied:
Thanks, @anton-vlasenko! Added in e96f28daa602f57322534a31e9acf88ad4948209.

anton-vlasenko commented on src/wp-includes/block-template-utils.php:
```diff
@@ -592,6 +592,19 @@ function _build_block_template_result_from_file( $template_file, $template_type
+			$template->origin      =
+				'theme' !== $template->origin && 'theme' !== $template->source ?
+				'plugin' :
+				$template->origin;
```
It appears that the expression does not modify the value of `$template->origin`. Could you take a look at line 587:

```php
$template->source = 'theme';
```

Since `$template->source` is always set to `'theme'`, the condition `'theme' !== $template->origin && 'theme' !== $template->source` always evaluates to `false`.

Aljullu replied:
Good catch. Yeah, I think it makes sense to remove this line. This is meant to set the correct `origin` of the template, but my understanding is that `_build_block_template_result_from_file()` is only used to retrieve theme files (L1180 and L1333-13336 of this same file), so in this flow we would never need to assign the `origin` to `plugin`.

I guess I messed it up when copying the changes from Gutenberg, as I had to use some filters there as I couldn't hook into these functions directly.

I removed this line in e9d85805aecf745122819b19909bb19478bd65e9.

Aljullu committed:
Add docblocks to tests

Aljullu committed:
Move WP_Block_Templates_Registry_Test under block-templates

anton-vlasenko commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1014,6 +1027,19 @@ function _build_block_template_result_from_post( $post ) {
 		}
 	}
 
+	if ( 'wp_template' === $post->template_type ) {
```
Did you mean
```suggestion
	if ( 'wp_template' === $post->post_type ) {
```
?
`WP_Post->template_type` is a dynamic property and is not declared on the `WP_Post` class.

anton-vlasenko commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+				$error_message,
+				'6.7.0'
+			);
+			return new WP_Error( $error_code, $error_message );
```
Return value is expected to be `'WP_Block_Template|false'`, `'WP_Error'` returned.

Aljullu replied:
True, that was a leftover from a previous refactor. Thanks for noticing! Fixed in 32bb1e2a5043e1af9754371de0245de7d0ae0827.

anton-vlasenko commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+	 *
+	 * @param string $template_name Template name including namespace.
+	 * @param array  $args          Optional. Array of template arguments.
+	 * @return WP_Block_Template|false The registered template on success, or false on failure.
```
This doesn't match the actual return type. This method may also return a `WP_Error` object.

anton-vlasenko commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+	 *
+	 * @since 6.7.0
+	 *
+	 * @return WP_Block_Template[]|false Associative array of `$template_name => $template` pairs.
```
This method doesn't seem to return a `false` value.
```suggestion
	 * @return WP_Block_Template[] Associative array of `$template_name => $template` pairs.
```

anton-vlasenko commented on src/wp-includes/class-wp-block-templates-registry.php:
```diff
@@ -0,0 +1,255 @@
+	 * @since 6.7.0
+	 *
+	 * @param string $template_name Template name including namespace.
+	 * @return WP_Block_Template|null|false The registered template, or null if it is not registered.
```
This method doesn't seem to return a `false` value.
```suggestion
	 * @return WP_Block_Template|null The registered template, or null if it is not registered.
```

anton-vlasenko commented on src/wp-includes/block-template.php:
```diff
@@ -358,3 +358,38 @@ function _resolve_template_for_new_post( $wp_query ) {
+ * Register a block template.
+ *
+ * @param string       $template_name  Template name in the form of `plugin_uri//template_name`.
+ * @param array|string $args           Object type or array of object types with which the taxonomy should be associated.
+ * @param array|string $args           {
```
There are 2 entries for the same `@args` parameter.

Aljullu replied:
Fixed in feaf88d1a3d3740b4e7c1c6ae716ba41e628ab20.

anton-vlasenko commented on src/wp-includes/block-template.php:
```diff
@@ -358,3 +358,38 @@ function _resolve_template_for_new_post( $wp_query ) {
+ * Unregister a block template.
+ *
+ * @param string $template_name Template name in the form of `plugin_uri//template_name`.
+ * @return true|WP_Error True on success, WP_Error on failure or if the template doesn't exist.
```
It seems this method returns `WP_Block_Template|WP_Error`.

Aljullu replied:
Fixed in 0252d6ff7708cde37b9c9fd6a8071c9dad5f4660.

noisysocks commented:
> Although this needs a little work before landing in the final release, I am a little tempted to merge the code in it's current form while performance is figured out so we can merge in the inital package update. What do you think @azaozz @noisysocks?

If there's no outstanding blocking comments I'm ok with this.

anton-vlasenko commented:
> > Although this needs a little work before landing in the final release, I am a little tempted to merge the code in it's current form while performance is figured out so we can merge in the inital package update. What do you think @azaozz @noisysocks?
> 
> If there's no outstanding blocking comments I'm ok with this.

@noisysocks Speaking of blocking comments, the code currently has some compatibility issues with PHP 8.2, as it introduces a dynamic class property (`WP_Block_Template::$plugin`).
I also have concerns regarding https://github.com/WordPress/wordpress-develop/pull/7125#discussion_r1742133430 (might not work as expected).
In my opinion, this PR may not be ready to be merged in its current state.

Aljullu committed:
Add  property to WP_Template object

Aljullu committed:
Fix wrong return type in register function

Aljullu committed:
Update src/wp-includes/block-template-utils.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

Aljullu committed:
Update src/wp-includes/class-wp-block-templates-registry.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

Aljullu committed:
Update src/wp-includes/class-wp-block-templates-registry.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

Aljullu committed:
Remove duplicate args param

Aljullu committed:
Fix wrong return type in unregister function

Aljullu committed:
Remove unnecessary reassignment of ->origin

Aljullu commented:
Thanks for the review @anton-vlasenko! I have addressed all your feedback, I would appreciate it if you can take another look.

There is also [this open discussion](https://github.com/WordPress/wordpress-develop/pull/7125/files/d843c4917a55e15cc34bbb61f246402c69699b1d#diff-d52bf95054b516a19c2a332fd0c85bdbab2cb8e5a25756a12b99128744262144) about finding a more performant way to retrieve the plugin author text, but I might need some assistance on how to proceed. Also I don't know if it's a blocker.

azaozz commented:
> Although this needs a little work before landing in the final release, I am a little tempted to merge the code in it's current form while performance is figured out

Sounds good to me too after all of @anton-vlasenko's suggestions and fixes were added. Would be nice to address/fix the performance improvements from above but as that affect only the site editor don't think it is a blocker.

peterwilsoncc committed:
Merge branch 'trunk' into add/plugin-template-registration-api

peterwilsoncc committed:
Rename test class.

peterwilsoncc commented:
I've merged trunk in and renamed a class in the tests folder (ie, it won't effect your source code). If @anton-vlasenko is happy with the change a few hours ago then I think it's good to go.

anton-vlasenko commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1014,6 +1023,19 @@ function _build_block_template_result_from_post( $post ) {
 	}
 
+	if ( 'wp_template' === $post->post_type ) {
+		$registered_template = WP_Block_Templates_Registry::get_instance()->get_by_slug( $post->slug );
```
`WP_Post::$slug` is a dynamic class property. The `WP_Post` class doesn't have a `$slug` property declared.
Did you mean to use `$template->slug`?

Aljullu replied:
Another good catch, thanks, @anton-vlasenko! Fixed in 6a6394d7b073500a2817c82f0181c8ba69e39c94.

anton-vlasenko commented on tests/phpunit/tests/block-templates/WpBlockTemplatesRegistry.php:
```diff
@@ -0,0 +1,270 @@
+/**
+ * Test WP_Block_Templates_Registry class.
+ *
+ * @covers WP_Block_Templates_Registry
```
https://docs.phpunit.de/en/10.5/annotations.html#coversdefaultclass

```suggestion
 * @coversDefaultClass WP_Block_Templates_Registry
```

anton-vlasenko reviewed (commented):
@Aljullu Please find another review below. Apologies for not including it with the previous one, as I just noticed it.

Aljullu committed:
Fix wrong object to get the slug from

Aljullu committed:
Update tests/phpunit/tests/block-templates/WpBlockTemplatesRegistry.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

Aljullu committed:
Fix wrong docblock comments

Aljullu committed:
Fix another wrong docblock comment

anton-vlasenko commented:
> I've merged trunk in and renamed a class in the tests folder (ie, it won't effect your source code). If @anton-vlasenko is happy with the change a few hours ago then I think it's good to go.

@peterwilsoncc I don't see any blockers at the moment. Thanks!

peterwilsoncc committed:
A few test nitpicks.

peterwilsoncc committed:
Merge branch 'trunk' into add/plugin-template-registration-api

peterwilsoncc committed:
Doc standards fixes.

peterwilsoncc commented:
I've merged in trunk and pushed a couple of test nitpicks in e8b2b2c1fa4047d0df2e1d4be624d3814fb971ff and some docs standards fixes in 9a18a250fe89552522188e2d9e2a254cf7e9dd1a

peterwilsoncc commented:
I think this is good to go in, @noisysocks are you able to take a look to make sure I am not missing something obvious that will effect the editor?

noisysocks commented:
Looks OK to me. We'll have to wait until packages are synced before we commit.

noisysocks commented:
Committed in r59073.
